sdk/forms/src/main/java/com/klaviyo/forms/KlaviyoWebViewDelegate.kt

```diff
@@ -15,129 +14,243 @@ import androidx.core.content.ContextCompat.startActivity
-        Registry.log.debug("Not loading into $it since its active")
-    } ?: run {
-        webView = buildWebView()
-        Registry.log.wtf("Loading html into $webView")
-        webView?.loadDataWithBaseURL(
-            Registry.config.baseCdnUrl,
-            html,
-            MIME_TYPE,
-            null,
-            null
-        )
-        timer?.cancel()
-        timer = Registry.clock.schedule(Registry.config.networkTimeout.toLong()) {
-            Registry.log.debug("IAF WebView Aborted: Timeout waiting for Klaviyo.js")
-            close()
+    /**
+     * When [WEB_MESSAGE_LISTENER] is supported, messages sent over the Native Bridge from JS are received here
+     */
+    override fun onPostMessage(
+        view: WebView,
+        message: WebMessageCompat,
+        sourceOrigin: Uri,
+        isMainFrame: Boolean,
+        replyProxy: JavaScriptReplyProxy
+    ) = message.data?.let { postMessage(it) } ?: Unit
+
+    /**
+     * When [WEB_MESSAGE_LISTENER] is NOT supported, messages sent over the Native Bridge from JS are received here
+     */
+    @android.webkit.JavascriptInterface
+    fun postMessage(message: String) = try {
+        Registry.log.debug("JS interface postMessage $message")
+        when (val bridgeMessage = BridgeMessage.decodeWebviewMessage(message)) {
+            BridgeMessage.HandShook -> handShook()
+            BridgeMessage.Show -> show()
+            is BridgeMessage.AggregateEventTracked -> createAggregateEvent(bridgeMessage)
+            is BridgeMessage.ProfileEvent -> createProfileEvent(bridgeMessage)
+            is BridgeMessage.DeepLink -> deepLink(bridgeMessage)
+            BridgeMessage.Close -> close()
+            is BridgeMessage.Abort -> abort(bridgeMessage.reason)
         }
-    }
+    } catch (e: Exception) {
+        Registry.log.warning("Failed to relay webview message: $message", e)
+    }.let { }
 
-    private fun handShook() {
-        Registry.log.info("Received message from JS bridge")
-        timer?.cancel()
-    }
+    /**
+     * Handle a [BridgeMessage.HandShook] message by clearing the timer
+     */
+    private fun handShook() = handshakeTimer?.cancel()
 
+    /**
+     * Handle a [BridgeMessage.Show] message by displaying the webview before the form animates in
+     */
     private fun show() = webView?.let { webView ->
         activity?.window?.decorView?.let { decorView ->
             decorView.post {
                 decorView.findViewById<ViewGroup>(android.R.id.content).addView(webView)
                 webView.visibility = View.VISIBLE
             }
         } ?: run {
-            Registry.log.error("Unable to show IAF - null activity context reference")
+            Registry.log.warning("Unable to show IAF - null activity reference")
         }
     } ?: run {
-        Registry.log.error("Unable to show IAF - null WebView reference")
+        Registry.log.warning("Unable to show IAF - null WebView reference")
     }
 
+    /**
```

Comment:
No substantive changes, I just grouped the `BridgeMessage` handling methods to the bottom and the webview delegate methods to the top.